ghost-constructors.md

```diff
@@ -0,0 +1,142 @@
+Ghost datatype constructors
+===========================
+
+This RFC proposes that each constructor of a datatype can be declared
+to be `ghost`. The proposal also applies to codatatypes, but, for
+brevity, the RFC will continue to say just "datatype". Declaring a
+constructor to be `ghost` limits usage of that constructor to ghost
+contexts and auto-initialization.  It also restricts discriminators
+and destructors in compiled contexts, so that it is never necessary at
+run time to check if a value is of a ghost variant.
+
+Syntax
+------
+
+The name of a constructor in a datatype declaration can be preceded
+with the keyword `ghost`.  This makes the constructor a ghost
+constructor. A value created by a ghost constructor is called a ghost
+variant, and a datatype that has one or more ghost constructors is
+called a ghost-variant datatype.
+
+Resolution
+----------
+
+For a ghost constructor `ghost G(...)`, the automatically
+generated discriminator `G?` is also ghost.  Discriminators for
+non-ghost constructors can be used in compiled contexts; however, use
+of these discriminators has a precondition that is checked by the
+verifier (explained below).
+
+In a `match` construct, a `case` pattern can be a variable that will
+bind to a particular part of the source expression. (If the name of
+this variable is irrelevant, the language allows it to be given as
+`_`.) In a compiled context, the type of such a variable is not
+allowed to be a ghost-variant datatype.
+
+A ghost-variant datatype is considered _not_ an equality-supporting
+type. That is, the type does not obtain the type characteristic
+`(==)`, and thus values of the type can be compared for equality only
+in ghost contexts.
```

Comment:
Is this an attempt to prevent ghost constructors from being used in any compiled expression? I worry that just the fact that you can successfully read such a value at runtime already allows control flow that shouldn't be, and that this construction will be unsound in a subtle way. It sounds like the following program would be valid and halt at runtime, but I would expect the `myArray[0]` reference to be rejected statically, if I am trying to soundly avoid ever reading uninitialized storage:

```dafny
datatype D = ghost G(x: int) | D0(y: int) | D1(z: int)

method Main() {
  var myArray := new D[5];
  var shouldNotReferenceThisAtRuntime := myArray[0];
  expect false;
}
```